### PR TITLE
math: Add Trig and Inv-Trig math functions

### PR DESCRIPTION
**Pull request by raoulb**
Add `Exp` and `Sin`/`Cos` functions to math for use in templates.

Rationale:
- There is already `Log` so also having `Exp` seems to be expected.
- `Sin`/`Cos` always comes in handy when dealing with rotations. (F.e. in context of the image processing functions.)

### CODE REVIEW

**Comment by bep** [comment]
This looks good and I guess useful to enough people to me. What's your take @jmooring ?

**Comment by jmooring** [comment]
If we're going to add trig functions, we should add these as well:
<https://github.com/gohugoio/hugo/issues/10857>

Also, given the infrequent use (this is only the second request I've seen), it would be better if these were not aliased.

**Comment by raoulb** [comment]
Ok, I'll update the PR with the other functions.

Should I remove the aliased variants?

**Review by phuchau1989 — APPROVED** [review]

File: docs/content/en/functions/math/ToDegrees.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+action:
+  aliases: []
+  related:
+    - functions/math/ToRadians
+    - functions/math/Pi
+  returnType: float64
+  signatures: [math.ToDegrees VALUE]
+---
+
+{{< new-in 0.129.0 >}}
```
Comment (jmooring) [inline docs/content/en/functions/math/ToDegrees.md:15]:
```suggestion
{{< new-in 0.130.0 >}}
```

File: docs/content/en/functions/math/ToRadians.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+action:
+  aliases: []
+  related:
+    - functions/math/ToDegrees
+    - functions/math/Pi
+  returnType: float64
+  signatures: [math.ToRadians VALUE]
+---
+
+{{< new-in 0.129.0 >}}
```
Comment (jmooring) [inline docs/content/en/functions/math/ToRadians.md:15]:
```suggestion
{{< new-in 0.130.0 >}}
```

File: docs/content/en/functions/math/ToRadians.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+```go-html-template
+{{ math.ToRadians 1.5707963267948966 }} → 90
```
Comment (jmooring) [inline docs/content/en/functions/math/ToRadians.md:18]:
```suggestion
{{ math.ToRadians 90 }} → 1.5707963267948966
```

File: tpl/math/math.go [hunk 4]
```diff
@@ -49,11 +49,51 @@ func (ns *Namespace) Abs(n any) (float64, error) {
 	return math.Abs(af), nil
 }
 
+// Acos returns the arccosine function of the number n.
```
Comment (jmooring) [inline tpl/math/math.go:52]:
```suggestion
// Acos returns the arccosine, in radians, of n.
```

File: tpl/math/math.go [hunk 5]
```diff
@@ -49,11 +49,51 @@ func (ns *Namespace) Abs(n any) (float64, error) {
...
+	}
+	return math.Acos(af), nil
+}
+
 // Add adds the multivalued addends n1 and n2 or more values.
 func (ns *Namespace) Add(inputs ...any) (any, error) {
 	return ns.doArithmetic(inputs, '+')
 }
 
+// Asin returns the arcsine function of the number n.
```
Comment (jmooring) [inline tpl/math/math.go:66]:
```suggestion
// Asin returns the arcsine, in radians, of n.
```

File: tpl/math/math.go [hunk 6]
```diff
@@ -49,11 +49,51 @@ func (ns *Namespace) Abs(n any) (float64, error) {
...
+func (ns *Namespace) Asin(n any) (float64, error) {
+	af, err := cast.ToFloat64E(n)
+	if err != nil {
+		return 0, errors.New("Asin operator can't be used with non integer or float value")
+	}
+	return math.Asin(af), nil
+}
+
+// Atan returns the arctangent function of the number n.
```
Comment (jmooring) [inline tpl/math/math.go:75]:
```suggestion
// Atan returns the arctangent, in radians, of n.
```

File: tpl/math/math.go [hunk 7]
```diff
@@ -49,11 +49,51 @@ func (ns *Namespace) Abs(n any) (float64, error) {
...
+func (ns *Namespace) Atan(n any) (float64, error) {
+	af, err := cast.ToFloat64E(n)
+	if err != nil {
+		return 0, errors.New("Atan operator can't be used with non integer or float value")
+	}
+	return math.Atan(af), nil
+}
+
+// Atan returns the arctangent function of the numbers n and m, determining the correct quadrant.
```
Comment (jmooring) [inline tpl/math/math.go:84]:
```suggestion
// Atan2 returns the arc tangent of n/m, using the signs of the two to determine the quadrant of the return value.
```

File: tpl/math/math.go [hunk 8]
```diff
@@ -64,6 +104,15 @@ func (ns *Namespace) Ceil(n any) (float64, error) {
 	return math.Ceil(xf), nil
 }
 
+// Cos returns the cosine function of the number n.
```
Comment (jmooring) [inline tpl/math/math.go:107]:
```suggestion
// Cos returns the cosine of the radian argument n.
```

File: tpl/math/math.go [hunk 9]
```diff
@@ -89,6 +138,40 @@ func (ns *Namespace) Log(n any) (float64, error) {
...
+func (ns *Namespace) Log10(n any) (float64, error) {
+	af, err := cast.ToFloat64E(n)
+	if err != nil {
+		return 0, errors.New("Log10 operator can't be used with non integer or float value")
+	}
+
+	return math.Log10(af), nil
+}
+
+// E returns the number constant E.
```
Comment (jmooring) [inline tpl/math/math.go:161]:
```suggestion
// Returns the mathematical constant e.
```

File: tpl/math/math.go [hunk 10]
```diff
@@ -89,6 +138,40 @@ func (ns *Namespace) Log(n any) (float64, error) {
...
+func (ns *Namespace) E() float64 {
+	return math.E
+}
+
+// Exp returns the exponential function of the number n.
```
Comment (jmooring) [inline tpl/math/math.go:166]:
```suggestion
// Exp returns the base-e exponential of n.
```

File: tpl/math/math.go [hunk 11]
```diff
@@ -146,6 +213,16 @@ func (ns *Namespace) Mul(inputs ...any) (any, error) {
 	return ns.doArithmetic(inputs, '*')
 }
 
+// Returns the golden ratio constant Phi.
```
Comment (jmooring) [inline tpl/math/math.go:216]:
```suggestion
// Returns the golden ratio constant phi.
```

File: tpl/math/math.go [hunk 12]
```diff
@@ -146,6 +213,16 @@ func (ns *Namespace) Mul(inputs ...any) (any, error) {
...
+func (ns *Namespace) Phi() float64 {
+	return math.Phi
+}
+
+// Pi returns the number constant pi.
```
Comment (jmooring) [inline tpl/math/math.go:221]:
```suggestion
// Pi returns the mathematical constant pi.
```

File: tpl/math/math.go [hunk 13]
```diff
@@ -173,6 +258,15 @@ func (ns *Namespace) Round(n any) (float64, error) {
 	return _round(xf), nil
 }
 
+// Sin returns the sine function of the number n.
```
Comment (jmooring) [inline tpl/math/math.go:261]:
```suggestion
// Sin returns the sine of the radian argument n.
```

File: tpl/math/math.go [hunk 14]
```diff
@@ -188,6 +282,43 @@ func (ns *Namespace) Sub(inputs ...any) (any, error) {
...
 
+// Sum returns the sum of all numbers in inputs. Any slices in inputs are flattened.
+func (ns *Namespace) Sum(inputs ...any) (sum float64, err error) {
+	fn := func(x, y float64) float64 {
+		return x + y
+	}
+	return ns.applyOpToScalarsOrSlices("Sum", fn, inputs...)
+}
+
+// Tan returns the tangent function of the number n.
```
Comment (jmooring) [inline tpl/math/math.go:293]:
```suggestion
// Tan returns the tangent of the radian argument n.
```

File: tpl/math/math.go [hunk 15]
```diff
@@ -49,11 +49,51 @@ func (ns *Namespace) Abs(n any) (float64, error) {
 	return math.Abs(af), nil
 }
 
+// Acos returns the arccosine function of the number n.
+func (ns *Namespace) Acos(n any) (float64, error) {
+	af, err := cast.ToFloat64E(n)
+	if err != nil {
+		return 0, errors.New("Acos operator can't be used with non integer or float value")
```
Comment (jmooring) [inline tpl/math/math.go:56]:
```suggestion
		return 0, errors.New("requires a numeric argument")
```

File: tpl/math/math.go [hunk 16]
```diff
@@ -49,11 +49,51 @@ func (ns *Namespace) Abs(n any) (float64, error) {
...
 // Add adds the multivalued addends n1 and n2 or more values.
 func (ns *Namespace) Add(inputs ...any) (any, error) {
 	return ns.doArithmetic(inputs, '+')
 }
 
+// Asin returns the arcsine function of the number n.
+func (ns *Namespace) Asin(n any) (float64, error) {
+	af, err := cast.ToFloat64E(n)
+	if err != nil {
+		return 0, errors.New("Asin operator can't be used with non integer or float value")
```
Comment (jmooring) [inline tpl/math/math.go:70]:
```suggestion
		return 0, errors.New("requires a numeric argument")
```

File: tpl/math/math.go [hunk 17]
```diff
@@ -49,11 +49,51 @@ func (ns *Namespace) Abs(n any) (float64, error) {
...
+	}
+	return math.Asin(af), nil
+}
+
+// Atan returns the arctangent function of the number n.
+func (ns *Namespace) Atan(n any) (float64, error) {
+	af, err := cast.ToFloat64E(n)
+	if err != nil {
+		return 0, errors.New("Atan operator can't be used with non integer or float value")
```
Comment (jmooring) [inline tpl/math/math.go:79]:
```suggestion
		return 0, errors.New("requires a numeric argument")
```

File: tpl/math/math.go [hunk 18]
```diff
@@ -49,11 +49,51 @@ func (ns *Namespace) Abs(n any) (float64, error) {
...
+	}
+	return math.Atan(af), nil
+}
+
+// Atan returns the arctangent function of the numbers n and m, determining the correct quadrant.
+func (ns *Namespace) Atan2(n, m any) (float64, error) {
+	afx, err := cast.ToFloat64E(n)
+	if err != nil {
+		return 0, errors.New("Atan2 operator can't be used with non integer or float value")
```
Comment (jmooring) [inline tpl/math/math.go:88]:
```suggestion
		return 0, errors.New("requires a numeric argument")
```

File: tpl/math/math.go [hunk 19]
```diff
@@ -49,11 +49,51 @@ func (ns *Namespace) Abs(n any) (float64, error) {
...
+	}
+	afy, err := cast.ToFloat64E(m)
+	if err != nil {
+		return 0, errors.New("Atan2 operator can't be used with non integer or float value")
```
Comment (jmooring) [inline tpl/math/math.go:92]:
```suggestion
		return 0, errors.New("requires a numeric argument")
```

File: tpl/math/math.go [hunk 20]
```diff
@@ -64,6 +104,15 @@ func (ns *Namespace) Ceil(n any) (float64, error) {
 	return math.Ceil(xf), nil
 }
 
+// Cos returns the cosine function of the number n.
+func (ns *Namespace) Cos(n any) (float64, error) {
+	af, err := cast.ToFloat64E(n)
+	if err != nil {
+		return 0, errors.New("Cos operator can't be used with non integer or float value")
```
Comment (jmooring) [inline tpl/math/math.go:111]:
```suggestion
		return 0, errors.New("requires a numeric argument")
```

File: tpl/math/math.go [hunk 21]
```diff
@@ -89,6 +138,40 @@ func (ns *Namespace) Log(n any) (float64, error) {
 	return math.Log(af), nil
 }
 
+// Log2 returns the binary logarithm of the number n.
+func (ns *Namespace) Log2(n any) (float64, error) {
+	af, err := cast.ToFloat64E(n)
+	if err != nil {
+		return 0, errors.New("Log2 operator can't be used with non integer or float value")
```
Comment (jmooring) [inline tpl/math/math.go:145]:
```suggestion
		return 0, errors.New("requires a numeric argument")
```

File: tpl/math/math.go [hunk 22]
```diff
@@ -89,6 +138,40 @@ func (ns *Namespace) Log(n any) (float64, error) {
...
+	}
+
+	return math.Log2(af), nil
+}
+
+// Log10 returns the decimal logarithm of the number n.
+func (ns *Namespace) Log10(n any) (float64, error) {
+	af, err := cast.ToFloat64E(n)
+	if err != nil {
+		return 0, errors.New("Log10 operator can't be used with non integer or float value")
```
Comment (jmooring) [inline tpl/math/math.go:155]:
```suggestion
		return 0, errors.New("requires a numeric argument")
```

File: tpl/math/math.go [hunk 23]
```diff
@@ -89,6 +138,40 @@ func (ns *Namespace) Log(n any) (float64, error) {
...
+// E returns the number constant E.
+func (ns *Namespace) E() float64 {
+	return math.E
+}
+
+// Exp returns the exponential function of the number n.
+func (ns *Namespace) Exp(n any) (float64, error) {
+	af, err := cast.ToFloat64E(n)
+	if err != nil {
+		return 0, errors.New("Exp operator can't be used with non integer or float value")
```
Comment (jmooring) [inline tpl/math/math.go:170]:
```suggestion
		return 0, errors.New("requires a numeric argument")
```

File: tpl/math/math.go [hunk 24]
```diff
@@ -173,6 +258,15 @@ func (ns *Namespace) Round(n any) (float64, error) {
 	return _round(xf), nil
 }
 
+// Sin returns the sine function of the number n.
+func (ns *Namespace) Sin(n any) (float64, error) {
+	af, err := cast.ToFloat64E(n)
+	if err != nil {
+		return 0, errors.New("Sin operator can't be used with non integer or float value")
```
Comment (jmooring) [inline tpl/math/math.go:265]:
```suggestion
		return 0, errors.New("requires a numeric argument")
```

File: tpl/math/math.go [hunk 25]
```diff
@@ -188,6 +282,43 @@ func (ns *Namespace) Sub(inputs ...any) (any, error) {
...
+		return x + y
+	}
+	return ns.applyOpToScalarsOrSlices("Sum", fn, inputs...)
+}
+
+// Tan returns the tangent function of the number n.
+func (ns *Namespace) Tan(n any) (float64, error) {
+	af, err := cast.ToFloat64E(n)
+	if err != nil {
+		return 0, errors.New("Tan operator can't be used with non integer or float value")
```
Comment (jmooring) [inline tpl/math/math.go:297]:
```suggestion
		return 0, errors.New("requires a numeric argument")
```

File: tpl/math/math.go [hunk 26]
```diff
@@ -188,6 +282,43 @@ func (ns *Namespace) Sub(inputs ...any) (any, error) {
...
+	}
+	return math.Tan(af), nil
+}
+
+// ToDegrees converts radians into degrees.
+func (ns *Namespace) ToDegrees(n any) (float64, error) {
+	af, err := cast.ToFloat64E(n)
+	if err != nil {
+		return 0, errors.New("the math.ToDegrees function requires a numeric argument")
```
Comment (jmooring) [inline tpl/math/math.go:306]:
```suggestion
		return 0, errors.New("requires a numeric argument")
```

File: tpl/math/math.go [hunk 27]
```diff
@@ -188,6 +282,43 @@ func (ns *Namespace) Sub(inputs ...any) (any, error) {
...
+	}
+
+	return af * 180 / math.Pi, nil
+}
+
+// ToRadians converts degrees into radians.
+func (ns *Namespace) ToRadians(n any) (float64, error) {
+	af, err := cast.ToFloat64E(n)
+	if err != nil {
+		return 0, errors.New("the math.ToRadians function requires a numeric argument")
```
Comment (jmooring) [inline tpl/math/math.go:316]:
```suggestion
		return 0, errors.New("requires a numeric argument")
```

**Review by jmooring — CHANGES_REQUESTED** [review]
Once you made the changes and compiled, run "hugo gen docshelper" to update the data.yaml file.

**Comment by bep** [comment]
@raoulb A general remark: I will squash all the commit into one when merging (I will make sure the commit message looks good), so if you push the "commit suggestion" buttons above, everything will be fine in the end with the correct attributions.

**Comment by raoulb** [comment]
> Once you made the changes and compiled, run "hugo gen docshelper" to update the data.yaml file.

Thanks, I did not know about the docshelper :D

**Comment by raoulb** [comment]
> @raoulb A general remark: I will squash all the commit into one when merging (I will make sure the commit message looks good), so if you push the "commit suggestion" buttons above, everything will be fine in the end with the correct attributions.

I did some more rebase and squash. Feel free to merge this all in one single commit if you like.

**Comment by raoulb** [comment]
Thanks for the review. I think I fixed everything now. Sorry for all the copy-paste errors :-/

**Comment by jmooring** [comment]
Please remove `math.Cbrt` and all references to it. If someone actually needs to perform this calculation they can use `math.Pow` with a fractional exponent. As a general note, adding more functionality in the middle of a review cycle is annoying.

**Comment by raoulb** [comment]
> Please remove `math.Cbrt` and all references to it.

Done.

> If someone actually needs to perform this calculation they can use `math.Pow` with a fractional exponent. 

True, but this also goes for `Sqrt` and some other things. I like a batteries-included 
concept. Scoped in the appropriate namespace it doesn't really have a cost impact for Hugo.

> As a general note, adding more functionality in the middle of a review cycle is annoying.

Sorry about that.

**Comment by bep** [comment]
OK, we need to touch base on the scope of this PR. 

This started out as "Exp and Sin/Cos".

Then the funcs from #10857 was added:

```
math.Acos
math.Asin
math.Atan
math.Cos
math.Sin
math.Tan
math.Pi
```


That's 8 funcs. I count 16 in this PR. Where did the additional funcs come from and are they really must haves? /cc @jmooring

**Comment by raoulb** [comment]
Yes, there are some more functions than initially planned. As I started wrapping the math library I found some more things that I think were missing.

It's about a relatively complete support for the Exp/Log pairs and the Trig/InvTrig pairs and some related helpers mentioned in the Issue/Forum post. For those I think it's just natural to expect to have them all. And I added Phi which I can imagine be useful for layout purposes.

Are they all really a *must have*? Depends, maybe, maybe not. But I'd rather like to have them if in doubt than miss them (and wait for an issues until someone needs them).

Again, as I already said: I'm all in favor of a batteries-included approach. In this case I don;t see any real downsides. All is properly put into the Math namespace, so does not clutter general scope for users. It's relatively little code that should require virtually no maintenance as it just wraps the Go std lib.

But if you disagree, fell free to cherry pick single commits (that's why it's still mostly single commits per topic) or I can split up the PR further and let some parts open for later.

**Comment by bep** [comment]
@raoulb OK, thanks for the explanation. It's incredibly hard to review a PR that's suddenly expanding/doubling in scope. What I suggest is this:

Reduce this PR to the 8 funcs listed above. Then create a PR #2 with the additional funcs.

**Comment by raoulb** [comment]
Yeah, maybe I should have stated to goal differently. I'd like to add some basic math functions to Hugo which are useful in general. I'll now split this PR into smaller chunks of code which implement functions that mathematically belong together.

**Comment by raoulb** [comment]
This PR now contains trig and inv-trig functions only. This closes issue https://github.com/gohugoio/hugo/issues/10857.

**Comment by raoulb** [comment]
I guess this is done now. Thanks for merging.

**Pull request closed without merging**